backend/modules/context_size_calculator.py

```
@@ -0,0 +1,204 @@
+                return AutoTokenizer.from_pretrained(local_path)
+            except Exception as e:
+                print(
+                    f"[ContextSizeCalculator] Error loading local tokenizer: {e}"
+                )
+
+        # Fall back to downloading from Hugging Face using download_tokenizer.py
+        print(
+            "[ContextSizeCalculator] Attempting to download tokenizer using download_tokenizer.py..."
+        )
+        success = self._download_tokenizer(hf_model_id)
+        if success:
+            # Try loading the newly downloaded tokenizer
+            try:
+                print(
+                    f"[ContextSizeCalculator] Loading newly downloaded tokenizer from: {local_path}"
+                )
+                return AutoTokenizer.from_pretrained(local_path)
+            except Exception as e:
+                print(
+                    f"[ContextSizeCalculator] Error loading newly downloaded tokenizer: {e}"
+                )
+
+        else:
+            print(
+                f"[ContextSizeCalculator] Failed to download tokenizer for {hf_model_id}"
+            )
+            return None
+
+    def _get_huggingface_model_id(self, model_id):
+        """
+        Map Ollama model ID to corresponding Hugging Face model ID
+        """
+        model_mapping = {
+            "mistral:7b-instruct-v0.3-q3_K_M": "mistralai/Mistral-7B-Instruct-v0.3",
+            "deepseek-coder:6.7b-instruct-q3_K_M": "deepseek-ai/deepseek-coder-6.7b-instruct",
+            "qwen2.5-coder:3b-instruct-q8_0": "Qwen/Qwen2.5-Coder-3B-Instruct",
+            "gemma3:4b-it-q4_K_M": "google/gemma-3-4b-it",
+            "phi4-mini:3.8b-q4_K_M": "microsoft/Phi-4-mini-instruct",
+            "tinyllama": "TinyLlama/TinyLlama-1.1B-Chat-v0.6",
+        }
+
+        # Try exact match first
+        if model_id in model_mapping:
+            return model_mapping[model_id]
+
+        # Default to a general tokenizer if we can't find a match
+        print(
```

Maybe raise a warning instead of print, if this info that we didn't find a tokenizer is important?